arangod/Agency/Agent.cpp

```diff
@@ -797,14 +848,80 @@ void Agent::advanceCommitIndex() {
       LOG_TOPIC("e24aa", DEBUG, Logger::AGENCY)
           << "Critical mass for commiting " << _commitIndex + 1 << " through "
           << index << " to read db, done";
-      // Wake up rest handlers:
+      // Wake up write rest handlers:
       _waitForCV.broadcast();
 
+      logsForTrigger();
+
       if (_commitIndex >= _state.nextCompactionAfter()) {
         _compactor.wakeUp();
       }
+
     }
   }
+
+}
+
+futures::Future<query_t> Agent::poll(
+  index_t const& index, double const& timeout) {
+
+  using namespace std::chrono;
+
+  index_t commitIndex;
+  std::vector<log_t> logs;
+  query_t readDB = nullptr;
+  {
+    READ_LOCKER(oLocker, _outputLock);
+    commitIndex = _commitIndex;
+    if (index == 0 || index < _state.firstIndex()) {  // deliver as if index = 0
+      readDB = std::make_shared<VPackBuilder>();
+      VPackArrayBuilder r(readDB.get());
+      _readDB.get().toBuilder(*readDB, true);
+    } else if (index <= commitIndex) {   // deliver immediately all logs since index
+      logs = _state.get(index, commitIndex);
+    }
+  }
+
+  if (!logs.empty() || readDB != nullptr) {
+    auto builder = std::make_shared<VPackBuilder>();
+    {
+      VPackObjectBuilder e(builder.get());
+      builder->add("accepted", VPackValue(true));
+      builder->add(VPackValue("result"));
+      VPackObjectBuilder r(builder.get());
+      builder->add("commitIndex", VPackValue(commitIndex));
+      if (!logs.empty()) {
+        builder->add("firstIndex", VPackValue(logs.front().index));
+        builder->add(VPackValue("log"));
+        VPackArrayBuilder ls(builder.get());
+        for (auto const& i : logs) {
+          VPackObjectBuilder l(builder.get());
+          builder->add("index", VPackValue(i.index));
+          builder->add("query", VPackSlice(i.entry->data()));
+        }
+      } else {
+        builder->add("firstIndex", VPackValue(0));
+        builder->add("readDB", readDB->slice()[0]);
+      }
+    }
+    return futures::makeFuture(std::move(builder));
+  }
+
+  std::lock_guard guard(_promLock);
+
+
+  auto tp = steady_clock::now() +
+    duration_cast<milliseconds>(duration<double>(timeout));
+  auto res = _promises.try_emplace(tp, futures::Promise<query_t>());
+
+  if (!res.second) {
+    THROW_ARANGO_EXCEPTION_MESSAGE(
+      TRI_ERROR_INTERNAL, "Failed to add promise for polling");
+  }
+  if (_lowestPromise > index) {
+    _lowestPromise = index;
```

Comment:
Here is a chance that we set `_lowestPromise` to 0, if `index` was 0. Is this intended?

Reply:
no, if index == 0, you never get here